### PR TITLE
chore: use setup.cfg as main config file

### PR DESCRIPTION
**Pull request by henryiii**
Followup to #817, related to #809.

---

I've split out the setup.cfg changes. I still would highly recommend setup.cfg, because:

* It is a static, easily parsable file, which is highly superior to a Python file - for example, cibuildwheel should soon gain support for reading the Requires-Python metadata from the PEP 621 location or from setup.cfg, but setup.py is impossible to reliably parse.
* It avoids many common issues, like opening files without with, encoding issues, and issues with accessing the version.
* It keeps setup.py as _just_ logic, making it clear what's "dynamic" and what's just regular metadata.
* It's actually readable, much less boilerplate
* Less chance of importing from distutils before setuptools
* The PyPA projects, like twine, build, etc, use setup.cfg.
* It's been supported since 2016
* It's really easy to just say the fields are dynamically specifiable as keywords to setup, while it is harder to say the keywords to setup might be in `[options]` or in `[metadata]` or in ...

While I'm really excited for PEP 621 support, that's quite some time away (looks like funding has not even been secured yet), and then there's another period for the tooling to mature and become common enough to be used, it won't ever support Python 2 / PyPy 2, etc, so I expect adoption to be slow. And setup.cfg is still not going away, with some tools strongly refusing (cough, flake8, cough cough) to ever add pyproject.toml support (MyPy is still setup.cfg only too).

If people are writing setup.cfg's, it will be much easier to transition to pyproject.toml than from setup.py.

### CODE REVIEW

**Comment by di** [comment]
Can we experiment with using https://pypi.org/project/sphinx-tabs/ or https://pypi.org/project/sphinxcontrib-contentui/ to provide both the `setup.py` and `setup.cfg` configuration options simultaneously?

**Comment by henryiii** [comment]
I went with exectuablebooks' sphinx-tabs, as I'm a Jupyterbook fan. (https://henryiii.github.io/level-up-your-python ;) ), and it likely is well maintained. Here's what the current state looks like:

Screen Shot 2021-01-29 at 2 44 00 PM

File: source/tutorials/packaging-projects.rst [hunk 1]
```diff
@@ -93,82 +94,183 @@ always include it. If you were to use a different build system, such as
...
 
+Creating setup.cfg and setup.py
+-------------------------------
+
+You can place the metadata for your project in one of two places. Static
+information should be placed in :file:`setup.cfg`. This is simpler, easier to
+read, and avoids many common errors, like encoding errors. Any items that are
+dynamic, as well as extension modules or extensions to setuptools, need to go
+into :file:`setup.py`. You can completely configure via :file:`setup.py` if you
+really want to; both methods are shown below.
```
Comment (di) [inline source/tutorials/packaging-projects.rst:106]: I think we can be a bit more opinionated here that static metadata should be preferred, and that dynamic metadata should be an escape hatch. We should also structure this in a way that we can easily add `pyproject.toml` once PEP 621 lands.

Something like:

---

There are two types of metadata: static and dynamic.

* Static metadata (`setup.cfg`): guaranteed to be the same every time. This is simpler, easier to
read, and avoids many common errors, like encoding errors.
* Dynamic metadata (`setup.py`): possibly non-deterministic. Any items that are dynamic or determined at install-time, as well as extension modules or extensions to setuptools, need to go into `setup.py`.

Static metadata should be preferred and dynamic metadata should be used only as an escape hatch when absolutely necessary.

File: source/tutorials/packaging-projects.rst [hunk 2]
```diff
@@ -93,82 +94,183 @@ always include it. If you were to use a different build system, such as
...
+
+.. tabs::
+
+  .. tab:: Configure with setup.cfg (static)
```
Comment (di) [inline source/tutorials/packaging-projects.rst:110]:
```suggestion
  .. tab:: setup.cfg (static)
```

A nit, but can we make `setup.cfg` monospaced here?

File: source/tutorials/packaging-projects.rst [hunk 3]
```diff
@@ -93,82 +94,183 @@ always include it. If you were to use a different build system, such as
...
+    :func:`setup()`; this enables computed values to be used. You will also need
+    :func:`setup()` for setting up extension modules for compilation.
+
+    .. code-block:: python
+
+        import setuptools
+
+        setuptools.setup()
+
+  .. tab:: Configure with setup.py (dynamic)
```
Comment (di) [inline source/tutorials/packaging-projects.rst:200]:
```suggestion
  .. tab:: setup.py (dynamic)
```

Same here about monospacing.

File: source/tutorials/packaging-projects.rst (hunk 2 above)
Comment (henryiii) [reply source/tutorials/packaging-projects.rst]: Don't think so, opened an issue.

**Review by di — APPROVED** [review]
LGTM, will give it a day or so to let other folks review as well before merging.

**Comment by pradyunsg** [comment]
May I suggest using [sphinx-inline-tabs](https://github.com/pradyunsg/sphinx-inline-tabs) instead? I think it looks nicer. :)

**Comment by di** [comment]
Oh, look at that! Very nice @pradyunsg. Does it support markup in the tab title?

**Comment by henryiii** [comment]
I'll try it soon! :)

**Comment by henryiii** [comment]
No, it doesn't, though it seems to try harder:

Screen Shot 2021-02-02 at 10 43 20 AM

**Comment by henryiii** [comment]
I'll push a version that uses this:

Screen Shot 2021-02-02 at 10 51 03 AM

**Review by gaborbernat — APPROVED** [review]

**Comment by gaborbernat** [comment]
@henryiii instead of posting the image it's better to post the link to the rendered document, e.g. https://python-packaging-user-guide--818.org.readthedocs.build/tutorials/packaging-projects/#configuring-metadata

**Comment by henryiii** [comment]
The link is available already, "Details" on the readthedocs check. Putting an image in shows what it looks like without having to click a link, and once I've changed it, you only see the new version with a link, so you can't compare. Though I could have made the image clickable, I suppose :)

**Comment by gaborbernat** [comment]
> The link is available already, "Details" on the readthedocs check. Putting an image in shows what it looks like without having to click a link, and once I've changed it, you only see the new version with a link, so you can't compare.

That link takes you to the main page, not the section you want to view the change rendered. It's nice to have a direct link available 😄

**Comment by di** [comment]
Thanks @henryiii!

**Comment by henryiii** [comment]
FYI, sphinx-tabs now supports rst in tab headings, including math. https://github.com/executablebooks/sphinx-tabs/pull/102

**Review by bfox85 — COMMENTED** [review]
V

**Comment by webknjaz** [comment]
Hey @bfox85, this sort of drive-through reviews is rather useless and is considered spam. Please use words if you have something meaningful to add here.

**Comment by pradyunsg** [comment]
> Does it support markup in the tab title?
> [snip]
> No, it doesn't, though it seems to try harder:

It does now!

https://github.com/pradyunsg/sphinx-inline-tabs/issues/13

**Pull request merged**